### PR TITLE
Skill testing harness

### LINKED ISSUE #2872: Implement CLI entry point and argument parsing (run.ts)
Main entry point that accepts:
  --agent <name> (required)
  --language <ts|rust> (required)
  --scenarios <dir> (required)
  --output <dir> (default: ./results)
  --scenario <name> (optional filter to run one scenario)
  --timeout <seconds> (global override)

Validates args, prints config summary, delegates to executor. No matrix iteration yet — single (agent, language) tuple.

### LINKED ISSUE #2873: Implement spec loader — parse YAML scenario files
Parse scenario YAML per design doc §7. For MVP:
- Load a single YAML file
- Validate top-level structure (scenario.name, scenario.steps)
- Resolve steps[].prompt as plain string (no language-conditional yet)
- Resolve steps[].expectedSkills as plain list
- Resolve steps[].verify (build only — no deploy/behavior yet)
- Substitute built-in template variables (${RUN_ID}, ${WORKSPACE})

Skip for MVP: language-conditional fields, matrix filtering, only_if/skip_if, behavior verification schema.

Tests: unit tests for spec loader with a fixture YAML.

### LINKED ISSUE #2874: Implement AgentDriver base interface and timeout handling
Define the AgentDriver interface per design doc §4:
- setup(workspace, skillsDir): Promise<void>
- sendPrompt(prompt, timeout): Promise<AgentResult>
- sendFollowup(prompt, timeout): Promise<AgentResult>
- teardown(): Promise<void>

Implement shared subprocess management:
- Spawn child process, capture stdout/stderr
- Enforce timeout via AbortController / kill signal
- Return AgentResult { success, output, durationSeconds, exitCode }

This is the base class, not a specific driver.

### LINKED ISSUE #2875: Implement filesystem watcher (Linux: inotifywait)
Per design doc §5.1–5.4:
- Start inotifywait monitoring skills directories
- Parse access log output to extract skill file reads
- Normalize paths to skill names
- Record baseline reads during agent startup
- After each subprompt, compute new reads since baseline
- Support cumulative activation tracking across steps

Linux-only for MVP. macOS (fswatch) deferred.

Test: create a temp dir, touch files, verify watcher detects reads.

### LINKED ISSUE #2876: Implement skill activation assertion engine
Per design doc §5.3:
- Assert expected skills ⊆ activated skills (default mode)
- Support allowedExtraSkills list
- Support strictSkillMatch mode (expected == activated exactly)
- Generate SKILL_NOT_ACTIVATED failure with actionable message

Takes watcher output + step spec, returns pass/fail with details.

### LINKED ISSUE #2878: Implement scenario executor — workspace lifecycle and step loop
Per design doc §8. Orchestrates a single scenario run.

Setup:
1. Create temp workspace directory
2. Symlink skills into workspace
3. Verify Golem connectivity (golem component list)
4. Start filesystem watcher
5. Initialize agent driver

Execution (for each step):
1. Record watcher baseline
2. Send prompt via agent driver (if prompt != null)
3. Check agent success
4. Assert skill activation
5. Run build verification (if verify.build)
6. Record step result

Teardown:
1. Stop watcher
2. Collect artifacts (logs, workspace snapshot)
3. Clean up Golem state
4. Remove workspace (if passed and cleanup: true)

This is the integration point. All other MVP components plug in here.

### LINKED ISSUE #2879: Implement JSON report generation (minimal)
Per design doc §9.2:
- Generate per-scenario JSON report with: scenario name, matrix tuple, run_id, status, duration, per-step results (status, duration, skills activated/expected, failure details if any), artifact paths
- Write to --output directory
- Print summary line to stdout (PASS/FAIL with step counts)

MVP: JSON only. HTML report deferred.

### LINKED ISSUE #2880: MVP integration test — full end-to-end run
Run the bootstrap scenario with all components wired together. Fix integration issues. Verify:
- Harness starts, loads spec, provisions workspace
- Claude Code receives prompt, generates project
- Filesystem watcher detects golem-new-project skill read
- golem build succeeds on generated project
- JSON report is written with PASS status
- Teardown cleans up workspace

This is a manual validation + bug-fix ticket, not a code ticket. Estimate accounts for integration debugging time.

### LINKED ISSUE #2881: Implement Claude Code driver
Concrete AgentDriver for Claude Code per design doc section 4.1. setup(): symlink skills into .claude/skills/ in workspace. sendPrompt(): run claude with --output-format json --max-turns 25 --cwd WORKSPACE. sendFollowup(): claude --resume --session-id SESSION. teardown(): clean up session state. Parse JSON output to extract success/failure and content. Handle edge cases: claude not installed, auth failure, rate limiting. Test: mock test that verifies CLI construction (no live agent needed).

### LINKED ISSUE #2882: Write bootstrap scenario: golem-new-project (TypeScript)
First scenario YAML to validate the harness end-to-end. Steps: 1. Prompt: Create a new Golem application called test-app with TypeScript. Expected skill: golem-new-project. Verify: build: true. This is intentionally minimal -- one step, one skill, build-only verification. The point is to prove the harness works. Note: the golem-new-project SKILL.md must exist before this scenario can run.

### LINKED ISSUE #2889: Implement deploy verification (Tier 2)
Per design doc section 6.2: Run golem deploy --yes in workspace. Assert exit code 0. Capture deploy output for failure reporting. Generate DEPLOY_FAILED with actionable message. Deploy implies build -- if verify.deploy is true, run build first.

### LINKED ISSUE #2895: Implement assertion engine — all assertion types
Per design doc section 6.4. Unified assertion evaluator supporting: stdout_contains, stdout_not_contains, stdout_matches, exit_code, status, body_contains, body_matches, result_json.path + result_json.equals. Each assertion produces a structured pass/fail result with expected vs. actual values for failure reporting.

### LINKED ISSUE #2897: Implement OpenCode driver (stub)
Stub driver for OpenCode. Reads both .claude/skills/ and .agents/skills/. Verify headless CLI mode exists before implementing. Mark as blocked if no headless mode available.

### LINKED ISSUE #2898: Implement Gemini CLI driver (stub)
Stub driver for Gemini CLI. Use gemini -p for non-interactive mode. Skill reading mechanism TBD -- may need custom solution. Mark as blocked until skill reading is confirmed.

### LINKED ISSUE #2906: CI step: Install and start Golem server
Download Golem binary for the runner OS/arch. Start golem server run --clean in background. Wait for server readiness (poll golem component list). Fail the job if server doesn't start within 60s.

### LINKED ISSUE #2907: CI step: Install coding agents (Claude Code, Codex)
Install the coding agent CLI for the matrix entry: claude-code: npm install -g @anthropic-ai/claude-code. codex: npm install -g @openai/codex. Configure API keys from GitHub Secrets. Verify installation: agent --version or equivalent.

### PR DESCRIPTION
**Pull request by Myestery**
## Summary
Merges the skill testing harness infrastructure into main. This includes the full agent testing harness for validating AI coding agents can discover and follow Golem skill files, CI workflow, and bootstrap scenarios.

## Issues
Closes #2872
Closes #2873
Closes #2874
Closes #2875
Closes #2876
Closes #2878
Closes #2879
Closes #2880
Closes #2881
Closes #2882
Closes #2898
Closes #2906
Closes #2907
Closes #2895
Closes #2897
Closes #2889
Closes https://github.com/golemcloud/golem/issues/2911, https://github.com/golemcloud/golem/issues/2916, https://github.com/golemcloud/golem/issues/2915, https://github.com/golemcloud/golem/issues/2913, https://github.com/golemcloud/golem/issues/2912, https://github.com/golemcloud/golem/issues/2903

Closes https://github.com/golemcloud/golem/issues/2885, closes https://github.com/golemcloud/golem/issues/2884, closes https://github.com/golemcloud/golem/issues/2914, closes https://github.com/golemcloud/golem/issues/2917, closes https://github.com/golemcloud/golem/issues/2904

### CODE REVIEW

File: golem-skills/skills/golem-new-project/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+---
```
Comment (vigoo) [inline golem-skills/skills/golem-new-project/SKILL.md:1]: Organization: 

let's move the whole thing into a subdirectory of the repo:

```
golem-skills/skills/...
golem-skills/tests/harness/...
```

so it is not interleaved with the server code
(similar to how we have `sdks/` etc)

File: tests/harness/src/run.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  const { agent, language, scenarios, output, scenario: scenarioFilter, timeout, skills: skillsDirRel, help } = values;
+
+  if (help || !agent || !language) {
+    const usage = `
+golem-skill-harness — Skill testing harness for Golem coding agents
+
+Usage:
+  npx tsx src/run.ts --agent <name> --language <ts|rust> --scenarios <dir> [options]
+
+Required:
```
Comment (vigoo) [inline tests/harness/src/run.ts]: I see how this is used in the CI with a matrix job, but we should make local runs more convenient. What I have in mind, let's say I write a new skill for all languages, I want to run it (locally) with all languages and coding agents to see if it's good enough before I submit a PR.

File: golem-skills/tests/harness/src/run.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+      process.exit(0);
+    }
+    console.error(chalk.red(usage));
+    process.exit(1);
+  }
+
+  const skillsDir = path.resolve(process.cwd(), skillsDirRel!);
+  const scenariosDir = path.resolve(process.cwd(), scenarios!);
+  const resultsDir = path.resolve(process.cwd(), output!);
+  const globalTimeoutSeconds = timeout ? Number.parseInt(timeout, 10) : undefined;
```
Comment (vigoo) [inline golem-skills/tests/harness/src/run.ts]: Is a 10 second timeout meaningful for tests like this? Reading the skill, actually using golem to do stuff, verifying. Isn't it taking minutes?

File: golem-skills/tests/harness/src/watcher.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+import * as fs from 'node:fs/promises';
+import * as path from 'node:path';
+
+export interface WatcherEvent {
+  path: string;
+  skillName: string;
+  timestamp: number;
+}
+
+export class SkillWatcher {
```
Comment (vigoo) [inline golem-skills/tests/harness/src/watcher.ts:11]: So did the filesystem watching proved to be the best way to detect if an agent uses skills? Did you investigate other ways as the design document proposed? Please write a summary of your findings

File: tests/harness/src/run.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  const scenarioReports: ScenarioReport[] = [];
+  let hasFailures = false;
+
+  for (const file of scenarioFiles) {
+    const spec = await ScenarioLoader.load(path.join(scenariosDir, file));
+
+    if (scenarioFilter && spec.name !== scenarioFilter) continue;
+
+    console.log(chalk.blue(`Running scenario: ${spec.name}`));
+    const workspace = path.join(process.cwd(), 'workspaces', spec.name.replace(/\s+/g, '-').toLowerCase());
```
Comment (vigoo) [inline tests/harness/src/run.ts]: Let's make sure this is `.gitignore`-d so it's not annoying when we run it locally

File: tests/harness/src/executor.ts [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+}).optional();
+
+const ScenarioSpecSchema = z.object({
+  name: z.string({ required_error: 'Scenario must have a "name" field' }),
+  settings: SettingsSchema,
+  prerequisites: PrerequisitesSchema,
+  steps: z.array(StepSpecSchema).min(1, 'Scenario must have at least one step'),
+});
+
+export interface StepSpec {
```
Comment (vigoo) [inline tests/harness/src/executor.ts]: The action a step describes (invoke/shell/trigger/prompt etc) should be a discriminated union - we can't have more than one in one step and this type currently does not reflect that

File: golem-skills/tests/harness/src/driver/claude.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
+import { BaseAgentDriver, AgentResult } from './base.js';
+
+export class ClaudeAgentDriver extends BaseAgentDriver {
```
Comment (vigoo) [inline golem-skills/tests/harness/src/driver/claude.ts:3]: Did you consider using ACP? Please write a summary of why not use it

File: tests/harness/scenarios/golem-new-project-ts.yaml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+name: "golem-new-project-ts"
+steps:
+  - id: "create-project"
+    prompt: "Create a new Golem application called test-app with TypeScript"
+    expectedSkills:
+      - "golem-new-project"
+    verify:
+      build: true
```
Comment (vigoo) [inline tests/harness/scenarios/golem-new-project-ts.yaml]: We should verify that it also deploys to have an end-to-end example, and also have as many of the steps and assertions as possible (even if not necessary for the final "new project skill") so we can see the test harness working.

File: tests/harness/src/run.ts [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    };
+
+    const reportPath = path.join(resultsDir, `${spec.name}.json`);
+    await fs.writeFile(reportPath, JSON.stringify(report, null, 2));
+    scenarioReports.push(report);
+
+    console.log(`${allPassed ? 'PASS' : 'FAIL'} ${spec.name} steps=${results.length}/${spec.steps.length}`);
+  }
+
+  // Aggregated summary report (#2912)
```
Comment (vigoo) [inline tests/harness/src/run.ts]: We should generate reports in [CTRF](https://ctrf.io) format so we can nicely integrate it in our CI flow
(this can be a separate ticket)

File: golem-skills/tests/harness/src/driver/claude.ts (hunk 7 above)
Comment (Myestery) [reply golem-skills/tests/harness/src/driver/claude.ts:3]: ACP is nice and well suited for creating interactive applications/editors that use multiple AI agents but I think its an overkill/mismatch for this use case.
Ideally what we want here is to be able to send prompts to AI agents, know when they are done, see the skill files they accessed and judge their output.
We also want to use it in a way that's closest to how the ideal user will use it.

Other things to consider
- ACP is made for editors
- ACP has no concept of skills use reporting
- ACP adds protocol overhead with close to zero benefit

File: golem-skills/tests/harness/src/watcher.ts (hunk 4 above)
Comment (Myestery) [reply golem-skills/tests/harness/src/watcher.ts:11]: I checked out ACP and also checked out some agents to see if there's an automated way to see skill usage from their output but it didn't work as planned. 
The other options I had was adding a "print skill usage" prompt to the agents and regex parsing their output.

File: golem-skills/tests/harness/src/run.ts (hunk 3 above)
Comment (Myestery) [reply golem-skills/tests/harness/src/run.ts]: I agree the default timeout isn't very meaningful. I will increase it to 120 cos that was what I was explicitly passing during testing.

File: tests/harness/src/run.ts (hunk 2 above)
Comment (Myestery) [reply tests/harness/src/run.ts]: It makes sense to make all defaults to target all options.

File: golem-skills/skills/golem-new-project/SKILL.md (hunk 1 above)
Comment (Myestery) [reply golem-skills/skills/golem-new-project/SKILL.md:1]: Done: moved everything under golem-skills/skills/ and golem-skills/tests/harness/. Updated CI workflow paths, .gitignore, and AGENTS.md

File: tests/harness/src/run.ts (hunk 2 above)
Comment (Myestery) [reply tests/harness/src/run.ts]: "Implemented : agent and --language now default to all. Running npx tsx src/run.ts with no flags runs every agent × language combination. Report filenames include {agent}-{language}-{scenario}.json to avoid collisions

File: golem-skills/tests/harness/src/run.ts (hunk 3 above)
Comment (Myestery) [reply golem-skills/tests/harness/src/run.ts]: Updated to 300 now

File: tests/harness/src/run.ts (hunk 5 above)
Comment (Myestery) [reply tests/harness/src/run.ts]: Yes this is already ignored

File: tests/harness/src/executor.ts (hunk 6 above)
Comment (Myestery) [reply tests/harness/src/executor.ts]: Done

File: tests/harness/scenarios/golem-new-project-ts.yaml (hunk 8 above)
Comment (Myestery) [reply tests/harness/scenarios/golem-new-project-ts.yaml]: Added the deploy step

File: tests/harness/src/run.ts (hunk 9 above)
Comment (Myestery) [reply tests/harness/src/run.ts]: tracking in https://github.com/golemcloud/golem/issues/2997 now

**Review by vigoo — APPROVED** [review]
Let's merge the other PRs based on this (when approved), then merge this.

**Comment by vigoo** [comment]
Before merge, unlink some issues that are not solved yet, if any (otherwise they will get closed automatically)
